### DISCUSSION TITLE
Failure in importing gomemif module

### DISCUSSION THREAD

**Discussion by nikhil-agarwal-git** [FDio/govpp#98] [Q&A]
Hi,

I am trying to build icmp_responder_cb.go (https://github.com/FDio/govpp/tree/master/extras/gomemif/examples/icmp_responder_cb) from outside the extras dir structure and facing following go build err. 

```
go: finding module for package go.fd.io/govpp/extras/gomemif/memif
go: downloading go.fd.io/govpp v0.7.0
github.com/gomemiftest imports
        go.fd.io/govpp/extras/gomemif/memif: module go.fd.io/govpp@latest found (v0.7.0), but does not contain package go.fd.io/govpp/extras/gomemif/memif
```

I basically intent to use gomemif for memif interfacing with VPP in my project. And for the same I was trying to some experiments with existing available example in gomemif package. 

On futher investigation, I realised that module name in go.mod present in extras directory does not match with the module github URL path. And if I update the module name to match that of github URL i.e (module go.fd.io/govpp/extras -> module github.com/nikhil-agarwal-git/govpp/extras ), it works fine.

Queries: 
1. Is gomemif production ready? Since it seems that it is recently added with govpp v0.7.0.
2. For the issue I am facing, is there any already known issue? By any chance, is this issue due to the recent movement of gomemif to govpp-github.

Also please suggest if the fix proposed would be ok or there is some gap in my understanding. 

Regards
Nikhil Agarwal

**Comment by ondrej-fabry** [comment]
The gomemif is part of submodule `go.fd.io/govpp/extras`.

> I am trying to build icmp_responder_cb.go ([master/extras/gomemif/examples/icmp_responder_cb](https://github.com/FDio/govpp/tree/master/extras/gomemif/examples/icmp_responder_cb)) from outside the extras dir structure and facing following go build err.

Outside how? Still in GoVPP module or your own Go module?

If you intend to use both GoVPP client for VPP API and gomemif, then you should be importing two separate Go modules in your `go.mod` file.

Run these commands to import the modules:

```shell
go get go.fd.io/govpp@latest
go get go.fd.io/govpp/extras@latest
```

Please share more of your code or how you are using the GoVPP & gomemif so we can better understand your workflow.

**Reply by nikhil-agarwal-git** [reply]
Hi @ondrej-fabry,  Thanks for your reply. Appologies for the deplayed reply. 

Basically we intend to use gomemif pkg in our go app for setting and using vpp memif. 
With that as a background requirement, I started exploring with the help of examples in gomemif i.e icmp_responder_cb.go.

I tried building this example as an independent app by moving it a separate dir say /home/ubuntu/test/

PWD = '/home/ubuntu/test/'

**$: ls -lrt** 
icmp_responder_cb.go

**$: go mod init icmp_responder_cb.go**
go: creating new go.mod: module icmp_responder_cb.go
go: to add module requirements and sums:
        go mod tidy

**$: go mod tidy**
go: finding module for package github.com/google/gopacket/layers
go: finding module for package github.com/google/gopacket
go: finding module for package github.com/pkg/profile
go: finding module for package go.fd.io/govpp/extras/gomemif/memif
go: downloading go.fd.io/govpp/extras v0.7.0
go: found github.com/google/gopacket in github.com/google/gopacket v1.1.19
go: found github.com/google/gopacket/layers in github.com/google/gopacket v1.1.19
go: found github.com/pkg/profile in github.com/pkg/profile v1.7.0
go: finding module for package go.fd.io/govpp/extras/gomemif/memif
icmp_responder_cb.go imports
        **go.fd.io/govpp/extras/gomemif/memif: module go.fd.io/govpp/extras@latest found (v0.7.0), but does not contain package go.fd.io/govpp/extras/gomemif/memif**


Above is the error which I am getting when I try to build "icmp_responder_cb.go" as an independent app. 


// from my previous post, I was able to make it work with following tweaks. 

> On futher investigation, I realised that module name in go.mod present in extras directory does not match with the module github URL path. And if I update the module name to match that of github URL i.e (module go.fd.io/govpp/extras -> module github.com/nikhil-agarwal-git/govpp/extras ), it works fine.


Hope this explains the issue and also please suggest if the fix proposed would be ok or is there some other work around.

Thanks
Nikhil

**Reply by ondrej-fabry** [reply]
@nikhil-agarwal-git check this relevant issue for context: https://github.com/FDio/govpp/issues/111

In short, running `GOPROXY=direct go mod tidy` should help.
